### PR TITLE
Update and sync softmax interface to modern style

### PR DESCRIPTION
**Pull request by elliottslaughter**
Similar to #1667, and NOT stacked on top of it. (I keep going back and forth, but the current version is independent.)

CPU kernel is still not implement. GPU kernel and test has been updated and should be in sync with task-spec now.

Properly implements `dim` to match PyTorch semantics, which as far as I can tell wasn't working previously.

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/flexflow-train/1668)

### CODE REVIEW

**Comment by lockshaw** [comment]
Merged in `06adfe419e1d7ebc01192e1f9a2c17935b374792`.
